File: DSA/LC/322_coin_change.py
from typing import List


class Solution1:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0

        for a in range(1, amount + 1):
            for c in coins:
                if a >= c:
                    dp[a] = min(dp[a], 1 + dp[a - c])
        return dp[amount] if dp[amount] != amount + 1 else -1


class Solution2:
    def coinChange(self, coins: List[int], amount: int) -> int:

        if amount == 0:
            return 0

        temp = dict()

        # Recursive solution
        res = []

        def helper(partial_sum: int):

            if partial_sum == 0:
                return 0
            if partial_sum < 0:
                return -1

            if partial_sum in temp:
                return temp[partial_sum]

            min_coins = float('inf')
            for coin in coins:
                res = helper(partial_sum - coin)
                if res >= 0 and res < min_coins:
                    min_coins = res + 1  # Because result is for partial_sum - coin

            temp[partial_sum] = min_coins if min_coins != float('inf') else -1
            return temp[partial_sum]

        return helper(amount)

File: DSA/LC/322_coin_change.cpp
#include <vector>
#include <limits>
#include <iostream>

class Solution
{
public:
    int coinChange(std::vector<int> &coins, int amount)
    {
        std::vector<int> dp(amount + 1, amount + 1);

        dp[0] = 0;

        for (int curr_price{}; curr_price <= amount; ++curr_price)
        {
            for (const auto coin : coins)
            {
                if (curr_price - coin >= 0 && dp[curr_price - coin] + 1 < dp[curr_price])
                {
                    dp[curr_price] = dp[curr_price - coin] + 1;
                }
            }
        }

        return dp[amount] == amount + 1 ? -1 : dp[amount];
    }
};

int main()
{
    Solution s{};
    std::vector<int> coins{1, 5, 10};
    auto a = s.coinChange(coins, 12);

    std::cout << a << "\n\n";
    return 0;
}